proxy-node-gateway/src/main/java/uk/gov/ida/notification/saml/validation/components/RequestIdWatcher.java

```diff
@@ -0,0 +1,27 @@
+package uk.gov.ida.notification.saml.validation.components;
+
+import java.util.concurrent.ConcurrentSkipListSet;
+
+import org.opensaml.saml.saml2.core.Response;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+
+import uk.gov.ida.notification.exceptions.hubresponse.InvalidHubResponseException;
+
+public class RequestIdWatcher {
+    private final ConcurrentSkipListSet<String> seenAuthnRequestIds = new ConcurrentSkipListSet<>();
+
+    public void observe(AuthnRequest authnRequest) {
+        seenAuthnRequestIds.add(authnRequest.getID());
+    }
+
+    public boolean haveSeenRequestFor(Response response) {
+        return seenAuthnRequestIds.contains(response.getInResponseTo());
+    }
+
+    public void validateSeenRequestFor(Response response) {
```

Comment:
Wouldn't we want to remove the request ID from the set once we've validated it's been seen? Or do we periodically clear the set? Maybe it's not an issue seen as this is only a temporary solution. Do we have a ticket in the backlog to find & implement a long-term solution?

Reply:
Yeah that might be a good plan actually. Would prevent replay attacks (somewhat).

Reply:
Talked to Chris. We'll fix this before going to integration so fine for now.

Reply:
If we remove the ID, it'll mean that if the user needs to repost the hub response to the gateway for any reason, they won't be able to. Is that an issue at all?
Simon mentioned potentially using a Guice cache, but it doesn't solve the issue about losing all the id's on reboot, or scaling.